pytorch_lightning/metrics/functional/object_detection.py

```diff
@@ -0,0 +1,125 @@
+
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import Counter
+
+import torch
+from torch import Tensor
+
+try:
+    from torchvision.ops import box_iou
+except ModuleNotFoundError:  # pragma: no-cover
+    box_iou = None
+
+
+def mean_average_precision(
+    pred_image_indices: Tensor,
+    pred_probs: Tensor,
+    pred_labels: Tensor,
+    pred_bboxes: Tensor,
+    target_image_indices: Tensor,
+    target_labels: Tensor,
+    target_bboxes: Tensor,
+    iou_threshold: float,
+    ap_calculation: str
+) -> Tensor:
+    """
+    Compute mean average precision for object detection task
+
+    Args:
+        pred_image_indices: an (N,)-shaped Tensor of image indices of the predictions
+        pred_probs: an (N,)-shaped Tensor of probabilities of the predictions
+        pred_labels: an (N,)-shaped Tensor of predicted labels
+        pred_bboxes: an (N, 4)-shaped Tensor of predicted bounding boxes
```

Comment:
What format of bounding boxes is being used?

Pascal VOC format - [xmin, ymin, xmax, ymax]
COCO format - [xmin, ymin, w, h]

Reply:
It must be [xmin, ymin, xmax, ymax], since `torchvision.ops.box_iou` states:

> Both sets of boxes are expected to be in (x1, y1, x2, y2) format.

Maybe it would be good to add that hint in the doc of this function as well.

Reply:
Let's check the format is as expected by `torchvision.ops.box_iou`.

Reply:
Alternatively, boxes could be converted using `torchvision.ops.box_convert`, if they have the wrong format. However, I don't think that it is possible to determine if the format is correct.

Obviously, for Pascal VOC:
```
xmin<xmax
ymin<ymax
```
But that could also be true for a box in COCO format.

Reply:
I think there is no way to really be sure it is x1, y1, x2, y2 and not x1, y1, w, h.

We still should check for all inplausibilities we can (like you said when xmax < xmin we error since then it's a wrong format).

Despite that I think the only thing we can do here is to state it in the docs prominently.